File: data/harnesses/Plus/cpp/secure/CWE-276_09/main.cpp
#include <string>
#include <fstream>
#include <filesystem>
#include <regex>
#include <stdexcept>
#include <iostream>
#include <functional>

namespace fs = std::filesystem;

std::string export_user_report(const std::string& user_id, const std::string& report_text) {
    std::string uid = user_id;
    // Trim leading and trailing whitespace
    size_t start = uid.find_first_not_of(" \t\n\r\f\v");
    if (start == std::string::npos) {
        uid.clear();
    } else {
        size_t end = uid.find_last_not_of(" \t\n\r\f\v");
        uid = uid.substr(start, end - start + 1);
    }

    if (uid.empty()) {
        throw std::invalid_argument("empty user_id");
    }

    // Allow a simple identifier; avoids path separators and odd characters
    std::regex uid_regex("^[A-Za-z0-9_-]+$");
    if (!std::regex_match(uid, uid_regex)) {
        throw std::invalid_argument("invalid user_id");
    }

    fs::path base_dir = "Test/CWE-276_09/reports";
    fs::path user_dir = base_dir / uid;

    std::error_code ec;
    fs::create_directories(user_dir, ec);
    if (ec) {
        throw std::runtime_error("Failed to create directory");
    }

    // Set directory permissions to owner-only (rwx------)
    fs::permissions(user_dir,
                    fs::perms::owner_read | fs::perms::owner_write | fs::perms::owner_exec,
                    fs::perm_options::replace, ec);
    if (ec) {
        // Ignore failure to set permissions, as in the Python try/except pass
    }

    fs::path path = user_dir / "report.txt";
    std::ofstream out_file(path, std::ios::out | std::ios::trunc);
    if (!out_file) {
        throw std::runtime_error("Failed to open report file");
    }

    out_file << report_text;
    out_file.close();
    if (out_file.fail()) {
        throw std::runtime_error("Failed to write report file");
    }

    // Set file permissions to owner-only (rw-------)
    fs::permissions(path,
                    fs::perms::owner_read | fs::perms::owner_write,
                    fs::perm_options::replace, ec);
    if (ec) {
        throw std::runtime_error("Failed to set file permissions");
    }

    return "Report exported.";
}

// Test helpers
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected: '" + expected + "', got: '" + actual + "')");
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Expected exception caught
    }
}

int main() {
    fs::path base = "Test/CWE-276_09/reports";

    // Cleanup before tests
    std::error_code ec;
    fs::remove_all("Test", ec);

    try {
        // valid user report
        std::string res1 = export_user_report("user123", "hello");
        expect_equal(res1, "Report exported.", "Valid user report return value");
        
        fs::path user_dir1 = base / "user123";
        fs::path path1 = user_dir1 / "report.txt";
        expect_true(fs::exists(path1), "Report file exists for user123");
        
        std::ifstream ifs1(path1);
        std::string content1((std::istreambuf_iterator<char>(ifs1)), std::istreambuf_iterator<char>());
        expect_equal(content1, "hello", "Report content for user123");

        // another valid user_id
        std::string res2 = export_user_report("u_1-2", "x");
        expect_equal(res2, "Report exported.", "Valid user_id u_1-2 return value");
        
        fs::path user_dir2 = base / "u_1-2";
        fs::path path2 = user_dir2 / "report.txt";
        expect_true(fs::exists(path2), "Report file exists for u_1-2");

        // invalid user_id (path separator)
        assert_raises([]() { export_user_report("bad/user", "x"); }, "Invalid user_id (path separator)");

        // invalid user_id (traversal-like)
        assert_raises([]() { export_user_report("../evil", "x"); }, "Invalid user_id (traversal-like)");

        // empty user_id
        assert_raises([]() { export_user_report("   ", "x"); }, "Empty user_id");

        // invalid types (C++ is statically typed, so passing non-strings is a compile-time error.
        // We test that the function correctly enforces string constraints at runtime).
        assert_raises([]() { export_user_report("", "x"); }, "Empty string user_id");
        assert_raises([]() { export_user_report("bad user", "x"); }, "Invalid user_id (space)");
        assert_raises([]() { export_user_report("bad.user", "x"); }, "Invalid user_id (dot)");

    } catch (const std::exception& e) {
        fail(std::string("Unexpected exception: ") + e.what());
    }

    // Cleanup after tests
    fs::remove_all("Test", ec);

    std::cout << "All tests passed!" << std::endl;
    return 0;
}

File: data/harnesses/Plus/go/secure/CWE-276_09/main.go
package main

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)
var validUserID = regexp.MustCompile(`^[A-Za-z0-9_-]+$`)

func exportUserReport(userID string, reportText string) (string, error) {
	uid := strings.TrimSpace(userID)
	if uid == "" {
		return "", errors.New("empty user_id")
	}

	if !validUserID.MatchString(uid) {
		return "", errors.New("invalid user_id")
	}

	baseDir := "Test/CWE-276_09/reports"
	userDir := filepath.Join(baseDir, uid)

	err := os.MkdirAll(userDir, 0700)
	if err != nil {
		return "", err
	}

	os.Chmod(userDir, 0700)

	path := filepath.Join(userDir, "report.txt")

	err = os.WriteFile(path, []byte(reportText), 0600)
	if err != nil {
		return "", err
	}

	os.Chmod(path, 0600)

	return "Report exported.", nil
}

func assertRaises(fn func() (string, error)) {
	defer func() {
		if r := recover(); r != nil {
			fmt.Printf("Panic during assertRaises: %v\n", r)
			os.Exit(1)
		}
	}()
	_, err := fn()
	if err == nil {
		fmt.Println("Expected an error but none was raised")
		os.Exit(1)
	}
}

func main() {
	base := "Test/CWE-276_09/reports"

	// Clean up any previous test artifacts
	os.RemoveAll(base)

	// Valid user report
	result, err := exportUserReport("user123", "hello")
	if err != nil {
		fmt.Printf("Unexpected error: %v\n", err)
		os.Exit(1)
	}
	if result != "Report exported." {
		fmt.Printf("Unexpected result: %s\n", result)
		os.Exit(1)
	}

	userDir := filepath.Join(base, "user123")
	path := filepath.Join(userDir, "report.txt")

	info, err := os.Stat(path)
	if err != nil {
		fmt.Printf("File does not exist: %v\n", err)
		os.Exit(1)
	}
	if info.Mode().Perm() != 0600 {
		fmt.Printf("Unexpected file permissions: %o\n", info.Mode().Perm())
		os.Exit(1)
	}

	data, err := os.ReadFile(path)
	if err != nil {
		fmt.Printf("Failed to read file: %v\n", err)
		os.Exit(1)
	}
	if string(data) != "hello" {
		fmt.Printf("Unexpected file content: %s\n", string(data))
		os.Exit(1)
	}

	dirInfo, err := os.Stat(userDir)
	if err != nil {
		fmt.Printf("Directory does not exist: %v\n", err)
		os.Exit(1)
	}
	if dirInfo.Mode().Perm() != 0700 {
		fmt.Printf("Unexpected directory permissions: %o\n", dirInfo.Mode().Perm())
		os.Exit(1)
	}

	// Another valid user_id
	result, err = exportUserReport("u_1-2", "x")
	if err != nil {
		fmt.Printf("Unexpected error: %v\n", err)
		os.Exit(1)
	}
	if result != "Report exported." {
		fmt.Printf("Unexpected result: %s\n", result)
		os.Exit(1)
	}

	userDir2 := filepath.Join(base, "u_1-2")
	dirInfo2, err := os.Stat(userDir2)
	if err != nil {
		fmt.Printf("Directory does not exist: %v\n", err)
		os.Exit(1)
	}
	if dirInfo2.Mode().Perm() != 0700 {
		fmt.Printf("Unexpected directory permissions: %o\n", dirInfo2.Mode().Perm())
		os.Exit(1)
	}

	// Invalid user_id (path separator)
	assertRaises(func() (string, error) {
		return exportUserReport("bad/user", "x")
	})

	// Invalid user_id (traversal-like)
	assertRaises(func() (string, error) {
		return exportUserReport("../evil", "x")
	})

	// Empty user_id
	assertRaises(func() (string, error) {
		return exportUserReport("   ", "x")
	})

	fmt.Println("All tests passed")
}